### ISSUE TITLE
Cap'n'proto 1.0 API changes book keeping

### ISSUE THREAD

**Issue by vlovich** [capnproto/capnproto#1312]
Let's document the set of backward-incompatible changes we want to make to cap'n'proto when a push for 1.0 is made:

[ ] explicit `end()` for `AsyncOutputStream` rather than assuming EOF on destruction.
[ ] Semantic change to what `newTee` limit means. AsyncTee should use a single shared buffer & the limit should apply to the size of that buffer. This also allow us to relax the `!=` in `TeeBranch::tryTee` to a `<`.

**Comment by kentonv** [comment]
> Semantic change to what newTee limit means. AsyncTee should use a single shared buffer & the limit should apply to the size of that buffer. This also allow us to relax the != in TeeBranch::tryTee to a <.

I think we can make that change without considering it a breaking change. Honestly I'd be surprised if anyone other than ourselves even use "tee" and even if they do, I'd be very surprised if anyone is actually broken by a change like this. We make changes with more impact than this pretty regularly.

The `end()` change, on the other hand, is something that forces all implementations of AsyncOutputStream and most callers to change.

**Comment by vlovich** [comment]
Oh ok. @harrishancock raised it as a more serious compat concern.

**Comment by vlovich** [comment]
Also, how do you want to manage the 1.0 change? Maybe a KJ_1_0 macro for the in-development branch?

While we're at it, perhaps we want to establish a minimum compiler version & standard language policy? C++11 is going on ~10 years old? Having a documented policy around support makes it easy to know when we can bump the version & simplify all the macro magic that's needed to keep that support around.

**Comment by kentonv** [comment]
@vlovich We currently require C++14, and really only promise support for the compiler versions that we test in CI. (Speaking of which, I think I need to update the docs as our tested compiler versions changed in ba54c267dbf7cf6a23f0878b34c7299f1b53825e.)

I'd be up for requiring C++17 or C++20 if there's a good reason. E.g. when we introduce async/await, I think we should just start requiring C++20 for kj-async and capnp-rpc (but maybe stay with C++14 for the serialization layer...). So far there hasn't been any other killer feature that seemed worth forcing all dependents to update their build beyond C++14.

> Maybe a KJ_1_0 macro for the in-development branch?

Nah we can just merge the changes without any guard. People who are tracking git had better be pinning to a commit, and can update than pin when they are ready.

**Comment by vlovich** [comment]
Yeah, requiring C++20 for kj-async I think may be a good idea to let us optimize KJ internals with async/await.

**Comment by Sarcasm** [comment]
FWIW: I'm a user of kj-async. My project uses KJ framework but not capnproto, and kj-async is used extensively.
I won't be able to use C++20 anytime soon, as my project roughly follows the VFX Reference Platform (https://vfxplatform.com/), right now at CY2020 (C++14) and C++17 in the foreseeable future.
I don't mean you should not support or require C++20, just that there is at least one user that would be stuck on pre-1.0.

**Comment by vlovich** [comment]
@Sarcasm is the `AsyncOutputStream` change important to you?

**Comment by Sarcasm** [comment]
I don't understand this change exactly but I don't see it as an issue to update to API changes, my project is the final product and its APIs do not expose the KJ dependency.
I have 2 simple implementations of `AsyncOutputStream`, and not sure about how many I use in total, but few.

**Comment by vlovich** [comment]
Gotcha. I don't think we'll move quickly on making kj-async require C++20. My preference would be to see if we can identify hotspots and maybe just have dual implementations until the coroutine ecosystem matures a bit.

**Comment by Sarcasm** [comment]
Optional C++20 support would be great.
For projects where I'm not limited to C++14/17 (personal projects) I would love to play with awaitable kj::Promise (watching https://github.com/capnproto/capnproto/pull/1104).

**Comment by vlovich** [comment]
Another one although this one is perhaps more controversial. `Promise` methods that consume the promise should require an r-value reference to make it more clear that the original variable is being left in an inconsistent state. That would be quite a lift to retrofit in KJ & Workers but I find myself constantly being nervous that I'm accessing the wrong thing when I have a consumption operation on an l-value reference.

**Comment by kentonv** [comment]
> Promise methods that consume the promise should require an r-value reference to make it more clear that the original variable is being left in an inconsistent state.

I've considered this before but I feel like the added `kj::mv`s are excessively ugly. I feel pretty comfortable with the idea that `Promise` is understood to be a linear type, where (almost) any method call consumes the promise.
